src/operator.rs

```diff
@@ -2,291 +2,484 @@ use std::fmt::Display;
+        None
+    }
+}
+pub fn apply_bit_shr_wrap(l: Num, r: Num) -> Option<Num> {
+    Some(l >> r)
+}
+pub fn apply_add(l: Num, r: Num) -> Option<Num> {
+    Some(l + r)
+}
+pub fn apply_sub(l: Num, r: Num) -> Option<Num> {
+    Some(l - r)
+}
+pub fn apply_mul(l: Num, r: Num) -> Option<Num> {
+    Some(l * r)
+}
+pub fn apply_mod_floor(l: Num, r: Num) -> Option<Num> {
+    if r == 0 || (Num::MIN < 0 && l == Num::MIN && r == !0) {
+        None
+    } else {
+        Some(num_integer::mod_floor(l, r))
+    }
+}
+pub fn apply_mod_trunc(l: Num, r: Num) -> Option<Num> {
+    l.checked_rem(r)
+}
+pub fn apply_div_floor(l: Num, r: Num) -> Option<Num> {
+    if r == 0 || (Num::MIN < 0 && l == Num::MIN && r == !0) {
+        None
+    } else {
+        Some(num_integer::div_floor(l, r))
     }
+}
+pub fn apply_div_trunc(l: Num, r: Num) -> Option<Num> {
+    l.checked_div(r)
+}
+pub fn apply_gcd(l: Num, r: Num) -> Option<Num> {
+    Some(num_integer::gcd(l, r))
+}
+pub fn apply_exp(l: Num, r: Num) -> Option<Num> {
+    l.checked_pow(r.try_into().ok()?)
+}
+pub fn apply_bit_neg(x: Num) -> Num {
+    !x
+}
+pub fn apply_neg(x: Num) -> Num {
+    0 - x
+}
 
-    #[inline]
-    pub fn vec_apply(&self, mut left: Vector, right: &Vector) -> Option<Vector> {
-        for (x, y) in left.iter_mut().zip(right.iter()) {
-            *x = self.apply(*x, *y)?;
-        }
-        Some(left)
+#[inline(always)]
+pub fn can_apply_unary_always(_: &Expr) -> bool {
+    true
+}
+#[inline(always)]
+pub fn can_apply_binary_always(_: &Expr, _: &Expr) -> bool {
+    true
+}
+#[inline(always)]
+pub fn can_apply_or(el: &Expr, er: &Expr) -> bool {
+    ok_before_keyword(el) && ok_after_keyword(er)
+}
+#[inline(always)]
+pub fn can_apply_space_or(el: &Expr, er: &Expr) -> bool {
+    !ok_before_keyword(el) && ok_after_keyword(er)
+}
+#[inline(always)]
+pub fn can_apply_or_space(el: &Expr, er: &Expr) -> bool {
+    ok_before_keyword(el) && !ok_after_keyword(er)
+}
+#[inline(always)]
+pub fn can_apply_space_or_space(el: &Expr, er: &Expr) -> bool {
+    !ok_before_keyword(el) && !ok_after_keyword(er)
+}
+
+pub const OP_OR: BinaryOp = BinaryOp {
+    name: "or",
+    prec: 3,
+    apply: apply_or,
+    can_apply: can_apply_or,
+    commutative: false,
+    right_assoc: false,
+};
+pub const OP_SPACE_OR: BinaryOp = BinaryOp {
+    name: " or",
+    prec: 3,
+    apply: apply_or,
+    can_apply: can_apply_space_or,
+    commutative: false,
+    right_assoc: false,
+};
+pub const OP_OR_SPACE: BinaryOp = BinaryOp {
+    name: "or ",
+    prec: 3,
+    apply: apply_or,
+    can_apply: can_apply_or_space,
+    commutative: false,
+    right_assoc: false,
+};
+pub const OP_SPACE_OR_SPACE: BinaryOp = BinaryOp {
+    name: " or ",
+    prec: 3,
+    apply: apply_or,
+    can_apply: can_apply_space_or_space,
+    commutative: false,
+    right_assoc: false,
+};
+pub const OP_LT: BinaryOp = BinaryOp {
+    name: "<",
+    prec: 5,
+    apply: apply_lt,
+    can_apply: can_apply_binary_always,
```

Comment:
We can either implement a Default trait for `BinaryOp`, or define a `OP_EMPTY` constant, and then we can do `..OP_EMPTY,` instead of having to specify `can_apply_binary_always` and the bool flags in every struct.